### PR TITLE
Feature/perceptron

### PR DESCRIPTION
**Pull request by Verntrax**

### CODE REVIEW

File: src/sample_sequence.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+
+class SampleSequence(utils.Sequence):
+    def __init__(self, sample_ids, sample_classes, batch_size):
+        self.ids, self.classes = sample_ids, sample_classes
+        self.batch_size = batch_size
+        self.cluster_db = h5py.File(config.clusters_db_path, 'r')
+
+    def __len__(self):
+        return int(np.ceil(len(self.ids)) / float(self.batch_size))
```
Comment (unitatem) [inline src/sample_sequence.py]: check yours parenthesis

File: src/perceptron_training.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import h5py
+from keras import metrics
+from keras import callbacks
+from src.sample_sequence import SampleSequence
+from keras.utils import to_categorical
+from keras.models import Sequential
+from keras.layers import Dense, Activation
+from src import config
+from random import shuffle
+import numpy as np
```
Comment (unitatem) [inline src/perceptron_training.py]: do you really need numpy ?

File: .gitignore [hunk 3]
```diff
@@ -8,4 +8,5 @@
 debug
 resources
 *.log
+src/__pycache__
```
[the remark this answers is not in the document]
Comment (Verntrax) [reply .gitignore]: Done

File: src/config.py [hunk 4]
```diff
@@ -1,5 +1,12 @@
-take_every_nth_sample = 4
-clusters_count = 8
+take_every_nth_sample = 1
+clusters_count = 20
+
+sizes_of_layers = [100]
+training_fraction = 0.75
+batch_size = 32
+epochs = 200
+min_delta = 0.01
```
[the remark this answers is not in the document]
Comment (Verntrax) [reply src/config.py]: Changed to min_improvement_required

File: src/config.py [hunk 5]
```diff
@@ -1,5 +1,12 @@
...
+patience = 2
```
[the remark this answers is not in the document]
Comment (Verntrax) [reply src/config.py]: Changed to max_no_improvement_epochs

File: src/config.py [hunk 6]
```diff
@@ -1,5 +1,12 @@
-take_every_nth_sample = 4
-clusters_count = 8
+take_every_nth_sample = 1
+clusters_count = 20
+
+sizes_of_layers = [100]
+training_fraction = 0.75
+batch_size = 32
+epochs = 200
```
[the remark this answers is not in the document]
Comment (Verntrax) [reply src/config.py]: Changed to max_epochs

File: src/config.py [hunk 7]
```diff
@@ -1,5 +1,12 @@
-take_every_nth_sample = 4
-clusters_count = 8
+take_every_nth_sample = 1
+clusters_count = 20
+
+sizes_of_layers = [100]
+training_fraction = 0.75
```
[the remark this answers is not in the document]
Comment (Verntrax) [reply src/config.py]: Done

File: src/perceptron_training.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+
+def create_class_mapping():
+    cluster_db = h5py.File(config.clusters_db_path, 'r')
+    class_names = list(cluster_db.keys())
+    sorted(class_names)
+    mapping = {}
+    for i, class_name in enumerate(class_names):
+        mapping[class_name] = i
+    cluster_db.close()
```
[the remark this answers is not in the document]
Comment (Verntrax) [reply src/perceptron_training.py]: Done

File: src/perceptron_training.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    for class_name in cluster_db:
+        class_ids = []
+        for photo_name in cluster_db[class_name]:
+            class_ids.append((class_name, photo_name))
+        shuffle(class_ids)
+        divide_point = int(len(class_ids) * config.training_fraction)
+        training_ids += class_ids[:divide_point]
+        test_ids += class_ids[divide_point:]
+        shuffle(training_ids)
+        shuffle(test_ids)
```
[the remark this answers is not in the document]
Comment (Verntrax) [reply src/perceptron_training.py]: Done

File: src/sample_sequence.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+    def __getitem__(self, idx):
+        batch_x = self.ids[idx * self.batch_size:(idx + 1) * self.batch_size]
```
[the remark this answers is not in the document]
Comment (Verntrax) [reply src/sample_sequence.py]: Done

File: src/sample_sequence.py (hunk 1 above)
Comment (Verntrax) [reply src/sample_sequence.py]: Done

File: src/perceptron_training.py (hunk 2 above)
Comment (Verntrax) [reply src/perceptron_training.py]: Deleted

File: src/perceptron_training.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    logging.info('Building perceptron: [' + ' '.join([str(x) for x in config.sizes_of_layers]) + ']')
+    model = Sequential()
+    for i, layer_size in enumerate(config.sizes_of_layers):
+        if i == 0:
+            model.add(Dense(layer_size, input_dim=config.clusters_count))
+        else:
+            model.add(Dense(layer_size))
+        model.add(Activation('sigmoid'))
+    cluster_db = h5py.File(config.clusters_db_path, 'r')
+    output_size = len(cluster_db)
```
[the remark this answers is not in the document]
Comment (Verntrax) [reply src/perceptron_training.py]: Done

File: src/perceptron_training.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Verntrax) [reply src/perceptron_training.py]: Done

**Pull request merged**